### PR TITLE
[PAY-936] Search users to chat screen

### PR DESCRIPTION
**Pull request by dharit-tan**
### Description
Adds a screen to search users to start chatting with.

Todo:
- Get loading spinner at bottom of list while fetching more users.

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

Android emulator.

Screenshot_1676606452

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

### CODE REVIEW

File: packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+type SearchUsersScreenProps = {
+  debounceMs?: number
+  defaultUserList?: {
+    userIds: ID[]
+    loadMore: () => void
+    loading: boolean
+  }
+}
+
+export const ChatUserListScreen = (props: SearchUsersScreenProps) => {
```
Comment (rickyrombo) [inline packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx:144]: `SearchUsersScreenProps` but `ChatUserListScreen`?

I think it's probably good to separate this component out to make it useable by other features in the future, and have one that's specific to Chats that renders each user with the chats options/actions. That's how I did it for web

File: packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+  const dispatch = useDispatch()
+
+  const { userIds, status } = useSelector(getUserList)
+  const users = useProxySelector(
+    (state) => {
+      const ids = hasQuery ? userIds : defaultUserList.userIds
+      const users = getUsers(state, { ids })
+      return ids.map((id) => users[id])
+    },
+    [hasQuery, userIds]
```
Comment (rickyrombo) [inline packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx:167]: I made some changes here recently on web - I think the proxy was to restrictive and wasn't getting mutuals on initial load. Not sure if relevant here!

File: packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+      dispatch(searchUsers({ query }))
+    } else {
+      defaultUserList.loadMore()
+    }
+  }, [hasQuery, query, status, defaultUserList, dispatch])
+
+  const handlePress = (item) =>
+    navigation.navigate('Chat', { chatId: item.chat_id })
+
+  const renderItem = ({ item: user, index }: { item: User; index: number }) => {
```
Comment (rickyrombo) [inline packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx:200]: I think this is the part we could pass in to keep the main screen generic

File: packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+          {users.length > 0 ? (
+            <FlatList
+              onEndReached={handleLoadMore}
+              data={users}
+              renderItem={renderItem}
+              keyExtractor={(user) => user.user_id}
+            />
+          ) : (
+            <LoadingSpinner style={styles.loadingSpinner} />
```
Comment (rickyrombo) [inline packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx:268]: I think I have this on web too, but this technically doesn't allow for an empty state. Something we'll probably want to ask about

**Review by rickyrombo — APPROVED** [review]

File: packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx (hunk 1 above)
Comment (dharit-tan) [reply packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx:144]: Agree. Gonna merge this for now and make it generic in a follow-up.

File: packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx (hunk 4 above)
Comment (dharit-tan) [reply packages/mobile/src/screens/chat-screen/ChatUserListScreen.tsx:268]: Yep good call.

**Pull request merged**